### PR TITLE
Mock stats repository

### PR DESCRIPTION
**Pull request by GRGool**
### **Introduce MockStatsRepository and Adapt StatsViewModel Tests**

---

### **Description**  
This PR introduces and integrates a new `MockStatsRepository` to streamline testing for `StatsViewModel`. The following updates are included:

---

### **Creation of `MockStatsRepository`**  

The `MockStatsRepository` serves as testable substitute for the actual `StatsRepository`. Key features include:

#### 1. **State Tracking**  
   - **Method Call Verification**:  
     - Tracks method calls using `trackCall` and stores them in an internal list.  
     - `wasMethodCalled(String)` verifies if a specific method was invoked.  
     - `getMethodCalls()` provides a complete list of invoked methods.  
   - **Argument Tracking**:  
     - Tracks arguments passed to methods through generic lambdas (`getStat` and `updateStat`).  

#### 2. **Simulating Success and Failure**  
   - Introduced a `shouldSucceed` flag to simulate repository behavior:  
     - On success (`shouldSucceed = true`), operations invoke success callbacks with expected data.  
     - On failure (`shouldSucceed = false`), failure callbacks are invoked with a predefined exception (`Exception("Simulated failure")`).  

#### 3. **Resetting State**  
   - The `reset()` method ensures the mock is reusable by:  
     - Resetting `shouldSucceed` to `true`.  
     - Clearing stored user stats.  
     - Resetting tracked method calls.  

#### 4. **Get and Update Operations**  
   - Provided generic helper methods to streamline operations:  
     - `getStat`: Retrieves stats for a given user and invokes the appropriate callback.  
     - `updateStat`: Updates stats for a user and ensures state changes occur only on success.  
   - These helpers simplify logic across all repository methods, ensuring consistency.

---

### **Testing the Mock**  

To validate the functionality of `MockStatsRepository`:  
   - Dedicated tests were written to ensure expected behavior for:  
     - `get` and `update` methods.  
     - Method call tracking and argument verification.  
   - Included parameterized tests to reduce redundancy.

---

### **Adapting `StatsViewModel` Tests**  

The `StatsViewModel` tests were updated to leverage the new `MockStatsRepository`:  
   - Verified that `StatsViewModel` correctly invokes repository methods and processes responses.  
   - Included parameterized tests to reduce redundancy.
   - Improved line coverage significantly.

---

### CODE REVIEW

File: app/src/test/java/com/github/se/signify/model/stats/StatsViewModelTest.kt [hunk 1]
```diff
@@ -2,136 +2,221 @@ package com.github.se.signify.model.stats
...
-    verify(statsRepositoryFirestore).updateHardExerciseStats(eq(userId), any(), any())
+    assertEquals(initialStats.lettersLearned + newLetter, statsViewModel.lettersLearned.value)
   }
 
   @Test
-  fun updateDailyQuestStatsShouldCallRepositoryUpdateDailyQuestStats() {
-    statsViewModel.updateDailyQuestStats()
-    verify(statsRepositoryFirestore).updateDailyQuestStats(eq(userId), any(), any())
-  }
+  fun updateLettersLearnedShouldHandleFailureCorrectly() {
```
Comment (nada-ezziti) [inline app/src/test/java/com/github/se/signify/model/stats/StatsViewModelTest.kt:208]: The updateLettersLearned tests do not account for duplicate letters or cases where the new letter is invalid (e.g., a non-alphabetic character). Suggestion: Add tests to handle such cases and verify that the repository either rejects invalid updates or ensures the integrity of the lettersLearned list.

**Review by nada-ezziti — APPROVED** [review]
**General Comments:**

**Code Quality:** The introduction of MockStatsRepository significantly improves testability by providing a clean and reusable mock. The tracking of method calls and argument verification is well-structured and provides robust testing capabilities. Additionally, the use of parameterized tests reduces redundancy and improves maintainability.


**Edge Cases:**  While the implementation covers many scenarios, there are some edge cases and improvements that could enhance the robustness of the mock and tests. I made some suggestions for this cases that can be implemented in the future.

File: app/src/test/java/com/github/se/signify/model/stats/MockStatsRepositoryTest.kt [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+  fun `reset should clear stats and method calls`() {
+    mockStatsRepository.getEasyExerciseStats(userId, {}, {})
+    mockStatsRepository.updateEasyExerciseStats(userId, {}, {})
+
+    mockStatsRepository.reset()
+
+    assertTrue(mockStatsRepository.getMethodCalls().isEmpty())
+    assertTrue(mockStatsRepository.wasMethodCalled("getEasyExerciseStats").not())
+    assertTrue(mockStatsRepository.wasMethodCalled("updateEasyExerciseStats").not())
+  }
```
[the remark this answers is not in the document]
Comment (GRGool) [reply app/src/test/java/com/github/se/signify/model/stats/MockStatsRepositoryTest.kt:151]: Thanks, will make sure to additionally test that the stat list is empty.

**Comment by GRGool** [comment]
Thanks for the review, as for the viewmodel tests my goal wasn't to extensively test it but only test basic functionality and callbacks, as well as improve line coverage. As you didn't request changes I assume this is fine for you.

File: app/src/test/java/com/github/se/signify/model/stats/MockStatsRepositoryTest.kt [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+
+  @Test
+  fun `getStat should return the correct value`() {
```
[the remark this answers is not in the document]
Comment (GRGool) [reply app/src/test/java/com/github/se/signify/model/stats/MockStatsRepositoryTest.kt:156]: I believe testing invalid or null user IDs is unnecessary for the mock since it's designed to simulate controlled responses. Should be ok to assume valid input here imo
